### DISCUSSION TITLE
Zigbee2Tasmota (Z2T) OTA firmware update for devices

### DISCUSSION THREAD

**Discussion by nagyrobi** [arendst/Tasmota#11173] [Ideas]
Is there a way to make OTA firmware updates of the Zigbee devices through Zigbee2Tasmota?
IKEA and LEDVANCE support this, moreover, some devices like IKEA's SYMFONISK Sound Controller have a hard time working properly with third party coordinators unless they get a recent firmware update. But it's a bit hard to perform an update without Ikea's original coordinator...

**Comment by s-hadinger** [comment]
No it's not implemented and will not be. The ESP8266 is too limited for this. The best option would be to have an external program driving the OTA update via TCP bridge, like ZHA is doing. But I don't know if ZHA does OTA upgrades.

**Reply by nagyrobi** [reply]
Agree ESP8266 is too limited to handle this on its own but it would be good to implement some helper commands in Tasmota which could trigger the operation of OTA pointing to some external app running on a PC which could then handle finalization of the procedure.

The problem is that if we switch to zha we loose tasmota zigbee...

**Reply by s-hadinger** [reply]
You can use a second device only for OTA, and then pair back to your main coordinator.

I have no other short term solution.

**Reply by yzg011** [reply]
put zigbee fw in the littefs
using ccloader to update fw
![ab40ec36d967ca5f0dfb205ac03b063](https://user-images.githubusercontent.com/24872854/110445141-cb8a1100-80f8-11eb-86d4-6bfe0c4dd86a.png)

**Reply by sfromis** [reply]
That looks like updating the firmware of the EZSP chip, not what was asked for, updating firmware of individual Zigbee devices like Ikea Symfonisk or whatever.

**Reply by songhellobulter** [reply]
> No it's not implemented and will not be. The ESP8266 is too limited for this. The best option would be to have an external program driving the OTA update via TCP bridge, like ZHA is doing. But I don't know if ZHA does OTA upgrades.

We have ESP32 for our z2m gateways and we can extend the flash. The gateways send data to the backend via MQTT. We need the OTA function at the gateway to update the Zigbee devices. Anyone has done this implementation? Thank you very much.

**Reply by sfromis** [reply]
If connecting the coordinator running Tasmota to Z2M (Zigbee2MQTT) via the TCP ser2net bridge feature of Tasmota, after that Tasmota is totally out of the loop with anything Zigbee, and updating end device firmware would be up to what you do with Z2M.

**Reply by songhellobulter** [reply]
OK. Thank you sfromis for your reply. So tasmota here is an interpreter and a bypass and the general OTA implementation at the ESP32 would work? I guess the coordinator won't work for every Zigbee devices if they use encryption with their brand coordinators?

**Reply by sfromis** [reply]
Tasmota is not even "interpreting" anything, raw serial data is wrapped for TCP traffic. You could of course upgrade Tasmota itself using the normal OTA implementation, but that makes zero difference to Zigbee or Zigbee end devices.

Connecting to Zigbee2MQTT means that Tasmota is no longer doing anything Zigbee, merely transporting bytes back and forth. The coordinator function should work for end devices following the standard Zigbee protocols, as supported by Z2M. In special cases, you may have to work with Z2M to get things working, as Tasmota is not involved in this.

This specifically includes anything to OTA upgrade firmware of Zigbee end devices through the Zigbee mesh network. That happens between Z2M and the end device, without Tasmota even "knowing" what happens.

**Reply by songhellobulter** [reply]
Sorry sfromis, I mistyped this: Our gateway is z2t (Zigbee to tasmota). At backend we also have a tasmota service and at the gateway it does transforming and also some interpreting. Any other suggestions. Thank you.

**Reply by s-hadinger** [reply]
> We have ESP32 for our z2m gateways and we can extend the flash. The gateways send data to the backend via MQTT. We need the OTA function at the gateway to update the Zigbee devices. Anyone has done this implementation? Thank you very much.

ESP32 has most certainly the memory needed to implement Zigbee OTA. But it's not at all in my priorities. If anyone wants to implement and test it, I can provide guidance and support

**Reply by songhellobulter** [reply]
Thank you s-hadinger. Yes I would like to try the implementation. I will have a look through first about the OTA of Zigbee devices. Then should I ask for help and guidance here? or would you like to suggest where I should start?

**Reply by s-hadinger** [reply]
Good news. I strongly suggest that you do this part in Berry using the existing `ZbSend` command. 

What makes it more complex to do, is that you need to manage a state machine to k'ow what to send next. State machine is far easier to implement in Berry

**Reply by tarontop** [reply]
Is there any latest progress? I have the same problem

**Reply by songhellobulter** [reply]
Sorry, I haven't managed to start this and we need this.
Is it possible s-hadinger  write a template for us? Thank you.

**Reply by s-hadinger** [reply]
What do you mean "a template". OTA through Zigbee is a whole state machine and a whole pile of code to write. Not just a template.

**Reply by songhellobulter** [reply]
I mean like a flowchart. Does the state machine deal with states like:

New OTA file
Zbsend
Block ACK
Send block
endOfOTA


Thank you.

**Reply by Jason2866** [reply]
Well if you really need this you should dive into this special zigbee topic. It is not a trivial feature. Zigbee is a standard, search the documentation and learn how it works. A simple flow chart will help you nothing. If you want or need it you have to do yourself. There is no wish list for Tasmota to get anything done.

A general rule there is no private email communication. Please delete your email contact in your post.

**Reply by songhellobulter** [reply]
Appreciated if you could provide as much guidance as possible. Where should we start? Here it is a simple OTA procedure. Where should we pay more attention? Thank you very much.
![image](https://github.com/user-attachments/assets/642d33f4-111a-4fa0-b215-0256d6265a0b)

**Reply by s-hadinger** [reply]
OTA is fully described in Zigbee ZCL specification chapter 11. There are 80 pages of specs, so it's not that simple. I haven't read them, I suggest you first read them, including the frame format of ZCL (chapter 2).

Scanning through the spec, it seems that you first need to send a "Image Notify" command (command 0x00) to the device on the OTA cluster. The payload is described in section 11.13.3.1.

**Comment by Jason2866** [comment]
This is updating the zigbee stack in the cc25xx. This is not new. The ccloader Tasmota driver to flash the cc25xx is already there and was ported from @Staars https://github.com/arendst/Tasmota/blob/development/tasmota/xdrv_46_ccloader.ino
@nagyrobi asked for a way to flash zigbee end devices via OTA with a new firmware.

**Reply by yzg011** [reply]
this way is not good， when using ccloader， then zigbee2tasmota can not use
